--- django/core/handlers/wsgi.py.old	2014-03-10 12:29:41.685632000 -0700
+++ django/core/handlers/wsgi.py	2014-03-10 12:29:26.885557000 -0700
@@ -1,3 +1,4 @@
+from threading import Lock
 from django.core.handlers.base import BaseHandler
 from django.core import signals
 from django.dispatch import dispatcher
@@ -175,13 +176,23 @@
     raw_post_data = property(_get_raw_post_data)
 
 class WSGIHandler(BaseHandler):
+    initLock = Lock()
+
     def __call__(self, environ, start_response):
         from django.conf import settings
 
         # Set up middleware if needed. We couldn't do this earlier, because
         # settings weren't available.
         if self._request_middleware is None:
-            self.load_middleware()
+            with self.initLock:
+                try:
+                    # Check that middleware is still uninitialized.
+                    if self._request_middleware is None:
+                        self.load_middleware()
+                except:
+                    # Unload whatever middleware we got
+                    self._request_middleware = None
+                    raise
 
         dispatcher.send(signal=signals.request_started)
         try:
