--- django/core/handlers/wsgi.py.bak	2014-01-14 17:56:33.146789000 -0800
+++ django/core/handlers/wsgi.py	2014-01-14 17:57:28.594921000 -0800
@@ -103,7 +103,7 @@
         return '%s%s' % (self.path, self.environ.get('QUERY_STRING', '') and ('?' + self.environ.get('QUERY_STRING', '')) or '')
 
     def is_secure(self):
-        return self.environ.has_key('HTTPS') and self.environ['HTTPS'] == 'on'
+        return 'wsgi.url_scheme' in self.environ and self.environ['wsgi.url_scheme'] == 'https'
 
     def _load_post_and_files(self):
         # Populates self._post and self._files
